mlflow/ml-package-versions.yml

```diff
@@ -31,6 +31,7 @@ pytorch:
     maximum: "1.9.0"
     requirements: ["torchvision", "scikit-learn"]
     run: |
+      python -c "import torch; print(torch.__version__)"
```

Comment:
This is interesting:

https://github.com/mlflow/mlflow/pull/4627/checks?check_run_id=3201138447#step:11:11

```
1.9.0+cu102
```

`torch.__version__` contains a local version identifier (`+cu102`), which I think should be stripped.

Reply:
`torchvision.__version__` also contains a local version identifier:

```
% python -c "import torchvision; print(torchvision.__version__)"
0.10.0+cu102
```

Reply:
We could do strip it or use `importlib_metadata` to get a package version from package metadata.

Reply:
|     | module                    | package                       | importlib_metadata.version   | __version__                                                                   |
|----:|:--------------------------|:------------------------------|:-----------------------------|:------------------------------------------------------------------------------|
|   0 | mlflow                    | mlflow                        | 1.19.1.dev0                  | 1.19.1.dev0                                                                   |
|   1 | mlflow                    | mlflow                        | 1.19.1.dev0                  | 1.19.1.dev0                                                                   |
|   2 | socks                     | PySocks                       | 1.7.1                        | 1.7.1                                                                         |
|   3 | sockshandler              | PySocks                       | 1.7.1                        | module 'sockshandler' has no attribute '__version__'                          |
|   4 | pickleshare               | pickleshare                   | 0.7.5                        | 0.7.5                                                                         |
|   5 | pygments                  | Pygments                      | 2.9.0                        | 2.9.0                                                                         |
|   6 | pexpect                   | pexpect                       | 4.8.0                        | 4.8.0                                                                         |
|   7 | pyparsing                 | pyparsing                     | 2.4.7                        | 2.4.7                                                                         |
|   8 | PIL                       | Pillow                        | 8.3.1                        | 8.3.1                                                                         |
|   9 | idna                      | idna                          | 3.2                          | 3.2                                                                           |
|  10 | packaging                 | packaging                     | 21.0                         | 21.0                                                                          |
|  11 | tensorboard               | tensorboard                   | 2.5.0                        | 2.5.0                                                                         |
|  12 | smart_open                | smart-open                    | 5.1.0                        | 5.1.0                                                                         |
|  13 | Misc                      | zipp                          | 3.5.0                        | module 'Misc' has no attribute '__version__'                                  |
|  14 | zipp                      | zipp                          | 3.5.0                        | module 'zipp' has no attribute '__version__'                                  |
|  15 | databricks_cli            | databricks-cli                | 0.14.3                       | module 'databricks_cli' has no attribute '__version__'                        |
|  16 | integration               | databricks-cli                | 0.14.3                       | module 'integration' has no attribute '__version__'                           |
|  17 | numexpr                   | numexpr                       | 2.7.3                        | 2.7.3                                                                         |
|  18 | pyasn1                    | pyasn1                        | 0.4.8                        | 0.4.8                                                                         |
|  19 | fastcore                  | fastcore                      | 1.3.21                       | 1.3.21                                                                        |
|  20 | ipython_genutils          | ipython-genutils              | 0.2.0                        | 0.2.0                                                                         |
|  21 | IPython                   | ipython                       | 7.25.0                       | 7.25.0                                                                        |
|  22 | h5py                      | h5py                          | 2.10.0                       | 2.10.0                                                                        |
|  23 | absl                      | absl-py                       | 0.13.0                       | module 'absl' has no attribute '__version__'                                  |
|  24 | astunparse                | astunparse                    | 1.6.3                        | 1.6.3                                                                         |
|  25 | jinja2                    | Jinja2                        | 3.0.1                        | 3.0.1                                                                         |
|  26 | traitlets                 | traitlets                     | 5.0.5                        | 5.0.5                                                                         |
|  27 | websocket                 | websocket-client              | 1.1.0                        | 1.1.0                                                                         |
|  28 | regex                     | regex                         | 2021.7.6                     | 2.5.96                                                                        |
|  29 | docutils                  | docutils                      | 0.16                         | 0.16                                                                          |
|  30 | oauthlib                  | oauthlib                      | 3.1.1                        | 3.1.1                                                                         |
|  31 | alembic                   | alembic                       | 1.4.1                        | 1.4.1                                                                         |
|  32 | torchvision               | torchvision                   | 0.10.0                       | 0.10.0+cu102                                                                  |
|  33 | grpc                      | grpcio                        | 1.32.0                       | 1.32.0                                                                        |
|  34 | six                       | six                           | 1.15.0                       | 1.16.0                                                                        |
|  35 | six                       | six                           | 1.15.0                       | 1.16.0                                                                        |
|  36 | keras_preprocessing       | Keras-Preprocessing           | 1.1.2                        | 1.1.2                                                                         |
|  37 | catboost                  | catboost                      | 0.26                         | 0.26                                                                          |
|  38 | preshed                   | preshed                       | 3.0.5                        | module 'preshed' has no attribute '__version__'                               |
|  39 | pathspec                  | pathspec                      | 0.9.0                        | 0.9.0                                                                         |
|  40 | google                    | google-auth                   | 1.33.1                       | module 'google' has no attribute '__version__'                                |
|  41 | google                    | protobuf                      | 3.17.3                       | module 'google' has no attribute '__version__'                                |
|  42 | scipy                     | scipy                         | 1.4.1                        | 1.4.1                                                                         |
|  43 | certifi                   | certifi                       | 2021.5.30                    | 2021.05.30                                                                    |
|  44 | certifi                   | certifi                       | 2021.5.30                    | 2021.05.30                                                                    |
|  45 | markupsafe                | MarkupSafe                    | 2.0.1                        | 2.0.1                                                                         |
|  46 | tqdm                      | tqdm                          | 4.61.2                       | 4.61.2                                                                        |
|  47 | py                        | py                            | 1.10.0                       | 1.10.0                                                                        |
|  48 | click                     | click                         | 7.1.2                        | 7.1.2                                                                         |
|  49 | matplotlib_inline         | matplotlib-inline             | 0.1.2                        | module 'matplotlib_inline' has no attribute '__version__'                     |
|  50 | pycodestyle               | pycodestyle                   | 2.7.0                        | 2.7.0                                                                         |
|  51 | toml                      | toml                          | 0.10.2                       | 0.10.2                                                                        |
|  52 | wheel                     | wheel                         | 0.36.2                       | 0.36.2                                                                        |
|  53 | wheel                     | wheel                         | 0.36.2                       | 0.36.2                                                                        |
|  54 | _yaml                     | PyYAML                        | 5.4.1                        | module '_yaml' has no attribute '__version__'                                 |
|  55 | yaml                      | PyYAML                        | 5.4.1                        | 5.4.1                                                                         |
|  56 | libxgboost                | xgboost                       | 1.4.2                        | No module named 'libxgboost'                                                  |
|  57 | xgboost                   | xgboost                       | 1.4.2                        | 1.4.2                                                                         |
|  58 | docker                    | docker                        | 5.0.0                        | 5.0.0                                                                         |
|  59 | sphinxcontrib             | sphinxcontrib-serializinghtml | 1.1.5                        | module 'sphinxcontrib' has no attribute '__version__'                         |
|  60 | sphinxcontrib             | sphinxcontrib-jsmath          | 1.0.1                        | module 'sphinxcontrib' has no attribute '__version__'                         |
|  61 | sphinxcontrib             | sphinxcontrib-htmlhelp        | 2.0.0                        | module 'sphinxcontrib' has no attribute '__version__'                         |
|  62 | sphinxcontrib             | sphinxcontrib-applehelp       | 1.0.2                        | module 'sphinxcontrib' has no attribute '__version__'                         |
|  63 | sphinxcontrib             | sphinxcontrib-devhelp         | 1.0.2                        | module 'sphinxcontrib' has no attribute '__version__'                         |
|  64 | sphinxcontrib             | sphinxcontrib-qthelp          | 1.0.3                        | module 'sphinxcontrib' has no attribute '__version__'                         |
|  65 | huggingface_hub           | huggingface-hub               | 0.0.12                       | 0.0.12                                                                        |
|  66 | cached_property           | cached-property               | 1.5.2                        | 1.5.2                                                                         |
|  67 | catalogue                 | catalogue                     | 2.0.4                        | module 'catalogue' has no attribute '__version__'                             |
|  68 | dateutil                  | python-dateutil               | 2.8.2                        | 2.8.2                                                                         |
|  69 | spacy                     | spacy                         | 3.1.1                        | 3.1.1                                                                         |
|  70 | soupsieve                 | soupsieve                     | 2.2.1                        | 2.2.1                                                                         |
|  71 | typing_extensions         | typing-extensions             | 3.7.4.3                      | module 'typing_extensions' has no attribute '__version__'                     |
|  72 | thinc                     | thinc                         | 8.0.8                        | 8.0.8                                                                         |
|  73 | sacremoses                | sacremoses                    | 0.0.45                       | 0.0.41                                                                        |
|  74 | wcwidth                   | wcwidth                       | 0.2.5                        | 0.2.5                                                                         |
|  75 | caffe2                    | torch                         | 1.9.0                        | module 'caffe2' has no attribute '__version__'                                |
|  76 | torch                     | torch                         | 1.9.0                        | 1.9.0+cu102                                                                   |
|  77 | _pytest                   | pytest                        | 3.2.1                        | 3.2.1                                                                         |
|  78 | pytest                    | pytest                        | 3.2.1                        | 3.2.1                                                                         |
|  79 | tensorboard_plugin_wit    | tensorboard-plugin-wit        | 1.8.0                        | module 'tensorboard_plugin_wit' has no attribute '__version__'                |
|  80 | alabaster                 | alabaster                     | 0.7.12                       | module 'alabaster' has no attribute '__version__'                             |
|  81 | imagesize                 | imagesize                     | 1.2.0                        | module 'imagesize' has no attribute '__version__'                             |
|  82 | cymem                     | cymem                         | 2.0.5                        | module 'cymem' has no attribute '__version__'                                 |
|  83 | parso                     | parso                         | 0.8.2                        | 0.8.2                                                                         |
|  84 | matplotlib                | matplotlib                    | 3.4.2                        | 3.4.2                                                                         |
|  85 | mpl_toolkits              | matplotlib                    | 3.4.2                        | module 'mpl_toolkits' has no attribute '__version__'                          |
|  86 | pylab                     | matplotlib                    | 3.4.2                        | module 'pylab' has no attribute '__version__'                                 |
|  87 | greenlet                  | greenlet                      | 1.1.0                        | 1.1.0                                                                         |
|  88 | tensorboard_data_server   | tensorboard-data-server       | 0.6.1                        | 0.6.1                                                                         |
|  89 | charset_normalizer        | charset-normalizer            | 2.0.3                        | 2.0.3                                                                         |
|  90 | _plotly_future_           | plotly                        | 5.1.0                        | module '_plotly_future_' has no attribute '__version__'                       |
|  91 | _plotly_utils             | plotly                        | 5.1.0                        | module '_plotly_utils' has no attribute '__version__'                         |
|  92 | jupyterlab_plotly         | plotly                        | 5.1.0                        | module 'jupyterlab_plotly' has no attribute '__version__'                     |
|  93 | plotly                    | plotly                        | 5.1.0                        | 5.1.0                                                                         |
|  94 | itsdangerous              | itsdangerous                  | 2.0.1                        | 2.0.1                                                                         |
|  95 | gunicorn                  | gunicorn                      | 20.1.0                       | 20.1.0                                                                        |
|  96 | decorator                 | decorator                     | 5.0.9                        | 5.0.9                                                                         |
|  97 | urllib3                   | urllib3                       | 1.26.6                       | 1.26.6                                                                        |
|  98 | mccabe                    | mccabe                        | 0.6.1                        | 0.6.1                                                                         |
|  99 | lightgbm                  | lightgbm                      | 3.2.1                        | 3.2.1                                                                         |
| 100 | requests                  | requests                      | 2.26.0                       | 2.26.0                                                                        |
| 101 | tabulate                  | tabulate                      | 0.8.9                        | 0.8.9                                                                         |
| 102 | bs4                       | beautifulsoup4                | 4.9.3                        | 4.9.3                                                                         |
| 103 | google_auth_oauthlib      | google-auth-oauthlib          | 0.4.4                        | module 'google_auth_oauthlib' has no attribute '__version__'                  |
| 104 | pyflakes                  | pyflakes                      | 2.3.1                        | 2.3.1                                                                         |
| 105 | wrapt                     | wrapt                         | 1.12.1                       | 1.12.1                                                                        |
| 106 | tokenizers                | tokenizers                    | 0.10.3                       | 0.10.3                                                                        |
| 107 | sklearn                   | scikit-learn                  | 0.24.2                       | 0.24.2                                                                        |
| 108 | tenacity                  | tenacity                      | 8.0.1                        | module 'tenacity' has no attribute '__version__'                              |
| 109 | sphinx                    | Sphinx                        | 3.5.4                        | 3.5.4                                                                         |
| 110 | snowballstemmer           | snowballstemmer               | 2.1.0                        | module 'snowballstemmer' has no attribute '__version__'                       |
| 111 | pasta                     | google-pasta                  | 0.2.0                        | module 'pasta' has no attribute '__version__'                                 |
| 112 | fastai                    | fastai                        | 1.0.61                       | 1.0.61                                                                        |
| 113 | rsa                       | rsa                           | 4.7.2                        | 4.7.2                                                                         |
| 114 | termcolor                 | termcolor                     | 1.1.0                        | module 'termcolor' has no attribute '__version__'                             |
| 115 | numpy                     | numpy                         | 1.18.5                       | 1.18.5                                                                        |
| 116 | werkzeug                  | Werkzeug                      | 2.0.1                        | 2.0.1                                                                         |
| 117 | gitdb                     | gitdb                         | 4.0.7                        | 4.0.7                                                                         |
| 118 | cachetools                | cachetools                    | 4.2.2                        | 4.2.2                                                                         |
| 119 | attr                      | attrs                         | 21.2.0                       | 21.2.0                                                                        |
| 120 | bottleneck                | Bottleneck                    | 1.3.2                        | 1.3.2                                                                         |
| 121 | cloudpickle               | cloudpickle                   | 1.6.0                        | 1.6.0                                                                         |
| 122 | babel                     | Babel                         | 2.9.1                        | 2.9.1                                                                         |
| 123 | joblib                    | joblib                        | 1.0.1                        | 1.0.1                                                                         |
| 124 | opt_einsum                | opt-einsum                    | 3.3.0                        | v3.3.0                                                                        |
| 125 | pytz                      | pytz                          | 2021.1                       | 2021.1                                                                        |
| 126 | graphviz                  | graphviz                      | 0.17                         | 0.17                                                                          |
| 127 | filelock                  | filelock                      | 3.0.12                       | 3.0.12                                                                        |
| 128 | appdirs                   | appdirs                       | 1.4.4                        | 1.4.4                                                                         |
| 129 | jedi                      | jedi                          | 0.18.0                       | 0.18.0                                                                        |
| 130 | sphinx_click              | sphinx-click                  | 3.0.1                        | module 'sphinx_click' has no attribute '__version__'                          |
| 131 | tensorflow_estimator      | tensorflow-estimator          | 2.3.0                        | module 'tensorflow_estimator' has no attribute '__version__'                  |
| 132 | smmap                     | smmap                         | 4.0.0                        | 4.0.0                                                                         |
| 133 | _brotli                   | brotlipy                      | 0.7.0                        | module 'brotli._brotli' has no attribute '__version__'                        |
| 134 | brotli                    | brotlipy                      | 0.7.0                        | module 'brotli' has no attribute '__version__'                                |
| 135 | selenium                  | selenium                      | 3.141.0                      | 3.141.0                                                                       |
| 136 | docs                      | Keras                         | 2.4.0                        | module 'docs' has no attribute '__version__'                                  |
| 137 | keras                     | Keras                         | 2.4.0                        | 2.4.0                                                                         |
| 138 | mako                      | Mako                          | 1.1.4                        | 1.1.4                                                                         |
| 139 | cycler                    | cycler                        | 0.10.0                       | 0.10.0                                                                        |
| 140 | kiwisolver                | kiwisolver                    | 1.3.1                        | 1.3.1                                                                         |
| 141 | pyasn1_modules            | pyasn1-modules                | 0.2.8                        | 0.2.8                                                                         |
| 142 | pydantic                  | pydantic                      | 1.8.2                        | module 'pydantic' has no attribute '__version__'                              |
| 143 | _distutils_hack           | setuptools                    | 52.0.0.post20210125          | module '_distutils_hack' has no attribute '__version__'                       |
| 144 | _distutils_hack           | setuptools                    | 52.0.0.post20210125          | module '_distutils_hack' has no attribute '__version__'                       |
| 145 | pkg_resources             | setuptools                    | 52.0.0.post20210125          | module 'pkg_resources' has no attribute '__version__'                         |
| 146 | pkg_resources             | setuptools                    | 52.0.0.post20210125          | module 'pkg_resources' has no attribute '__version__'                         |
| 147 | setuptools                | setuptools                    | 52.0.0.post20210125          | 57.4.0                                                                        |
| 148 | setuptools                | setuptools                    | 52.0.0.post20210125          | 57.4.0                                                                        |
| 149 | pandas                    | pandas                        | 1.3.1                        | 1.3.1                                                                         |
| 150 | tensorflow                | tensorflow                    | 2.3.0                        | 2.3.0                                                                         |
| 151 | markdown                  | Markdown                      | 3.3.4                        | 3.3.4                                                                         |
| 152 | _ast27                    | typed-ast                     | 1.4.3                        | No module named '_ast27'                                                      |
| 153 | _ast3                     | typed-ast                     | 1.4.3                        | No module named '_ast3'                                                       |
| 154 | typed_ast                 | typed-ast                     | 1.4.3                        | 1.4.3                                                                         |
| 155 | pip                       | pip                           | 21.1.3                       | 21.1.3                                                                        |
| 156 | spacy_legacy              | spacy-legacy                  | 3.0.8                        | module 'spacy_legacy' has no attribute '__version__'                          |
| 157 | flatbuffers               | flatbuffers                   | 1.12                         | module 'flatbuffers' has no attribute '__version__'                           |
| 158 | prometheus_client         | prometheus-client             | 0.11.0                       | module 'prometheus_client' has no attribute '__version__'                     |
| 159 | importlib_metadata        | importlib-metadata            | 4.6.1                        | module 'importlib_metadata' has no attribute '__version__'                    |
| 160 | blis                      | blis                          | 0.7.4                        | module 'blis' has no attribute '__version__'                                  |
| 161 | fastprogress              | fastprogress                  | 1.0.0                        | 0.2.7                                                                         |
| 162 | OpenSSL                   | pyOpenSSL                     | 20.0.1                       | 20.0.1                                                                        |
| 163 | keras_applications        | Keras-Applications            | 1.0.8                        | 1.0.8                                                                         |
| 164 | requests_oauthlib         | requests-oauthlib             | 1.3.0                        | 1.3.0                                                                         |
| 165 | murmurhash                | murmurhash                    | 1.0.5                        | module 'murmurhash' has no attribute '__version__'                            |
| 166 | prompt_toolkit            | prompt-toolkit                | 3.0.19                       | 3.0.19                                                                        |
| 167 | sqlalchemy                | SQLAlchemy                    | 1.4.22                       | 1.4.22                                                                        |
| 168 | _openssl                  | cryptography                  | 3.4.7                        | module 'cryptography.hazmat.bindings._openssl' has no attribute '__version__' |
| 169 | _padding                  | cryptography                  | 3.4.7                        | No module named '_padding'                                                    |
| 170 | cryptography              | cryptography                  | 3.4.7                        | 3.4.7                                                                         |
| 171 | git                       | GitPython                     | 3.1.18                       | 3.1.18                                                                        |
| 172 | _cffi_backend             | cffi                          | 1.14.6                       | 1.14.6                                                                        |
| 173 | cffi                      | cffi                          | 1.14.6                       | 1.14.6                                                                        |
| 174 | querystring_parser        | querystring-parser            | 1.2.4                        | module 'querystring_parser' has no attribute '__version__'                    |
| 175 | prometheus_flask_exporter | prometheus-flask-exporter     | 0.18.2                       | 0.18.2                                                                        |
| 176 | transformers              | transformers                  | 4.9.1                        | 4.9.1                                                                         |
| 177 | flake8                    | flake8                        | 3.9.2                        | 3.9.2                                                                         |
| 178 | _black_version            | black                         | 19.10b0                      | module '_black_version' has no attribute '__version__'                        |
| 179 | black                     | black                         | 19.10b0                      | 19.10b0                                                                       |
| 180 | blackd                    | black                         | 19.10b0                      | No module named 'aiohttp'                                                     |
| 181 | blib2to3                  | black                         | 19.10b0                      | module 'blib2to3' has no attribute '__version__'                              |
| 182 | flask                     | Flask                         | 2.0.1                        | 2.0.1                                                                         |
| 183 | editor                    | python-editor                 | 1.0.4                        | 1.0.4                                                                         |
| 184 | srsly                     | srsly                         | 2.4.1                        | 2.4.1                                                                         |
| 185 | pathy                     | pathy                         | 0.6.0                        | module 'pathy' has no attribute '__version__'                                 |
| 186 | nvidia_smi                | nvidia-ml-py3                 | 7.352.0                      | module 'nvidia_smi' has no attribute '__version__'                            |
| 187 | pynvml                    | nvidia-ml-py3                 | 7.352.0                      | module 'pynvml' has no attribute '__version__'                                |
| 188 | gast                      | gast                          | 0.3.3                        | module 'gast' has no attribute '__version__'                                  |
| 189 | pycparser                 | pycparser                     | 2.20                         | 2.20                                                                          |
| 190 | sqlparse                  | sqlparse                      | 0.4.1                        | 0.4.1                                                                         |
| 191 | wasabi                    | wasabi                        | 0.8.2                        | module 'wasabi' has no attribute '__version__'                                |

Reply:
- `importlib_metadata` is normally equal to `__version__`. `torch` is an exception.

Reply:
Got it! We need to merge https://github.com/mlflow/mlflow/pull/4632 into this PR and use it to determine whether or not each package is available on PyPI, right?

Reply:
Yep!